File: src/main/java/com/medcabinet/service/MedicineService.java
package com.medcabinet.service;

import com.medcabinet.model.Medicine;
import com.medcabinet.model.MedicineType;
import com.medcabinet.model.ApplicationArea;
import com.medcabinet.repository.MedicineRepository;
import com.medcabinet.repository.MedicineTypeRepository;
import com.medcabinet.repository.ApplicationAreaRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import java.time.LocalDate;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

@Service
public class MedicineService {

    @Autowired
    private MedicineRepository medicineRepository;

    @Autowired
    private MedicineTypeRepository medicineTypeRepository;

    @Autowired
    private ApplicationAreaRepository applicationAreaRepository;

    public List<Medicine> getAllMedicines() {
        return medicineRepository.findAll();
    }

    public Optional<Medicine> getMedicineById(Long id) {
        return medicineRepository.findById(id);
    }

    public Medicine createMedicine(Medicine medicine, Long typeId, Set<Long> areaIds) {
        MedicineType type = medicineTypeRepository.findById(typeId)
                .orElseThrow(() -> new RuntimeException("Medicine type not found"));

        Set<ApplicationArea> areas = areaIds.stream()
                .map(areaId -> applicationAreaRepository.findById(areaId)
                        .orElseThrow(() -> new RuntimeException("Application area not found")))
                .collect(Collectors.toSet());

        medicine.setType(type);
        medicine.setApplicationAreas(areas);

        return medicineRepository.save(medicine);
    }

    public Medicine updateMedicine(Long id, Medicine medicineDetails, Long typeId, Set<Long> areaIds) {
        Medicine medicine = medicineRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Medicine not found"));

        if (typeId != null) {
            MedicineType type = medicineTypeRepository.findById(typeId)
                    .orElseThrow(() -> new RuntimeException("Medicine type not found"));
            medicine.setType(type);
        }

        if (areaIds != null) {
            Set<ApplicationArea> areas = areaIds.stream()
                    .map(areaId -> applicationAreaRepository.findById(areaId)
                            .orElseThrow(() -> new RuntimeException("Application area not found")))
                    .collect(Collectors.toSet());
            medicine.setApplicationAreas(areas);
        }

        medicine.setName(medicineDetails.getName());
        medicine.setQuantity(medicineDetails.getQuantity());
        medicine.setUnit(medicineDetails.getUnit());
        medicine.setExpiryDate(medicineDetails.getExpiryDate());

        return medicineRepository.save(medicine);
    }

    public void deleteMedicine(Long id) {
        medicineRepository.deleteById(id);
    }

    public List<Medicine> searchMedicines(String name) {
        return medicineRepository.findByNameContainingIgnoreCase(name);
    }

    public List<Medicine> getExpiredMedicines() {
        return medicineRepository.findByExpiryDateBefore(LocalDate.now());
    }

    public List<Medicine> getExpiringSoonMedicines(int days) {
        return medicineRepository.findByExpiryDateBefore(LocalDate.now().plusDays(days));
    }

    public List<Medicine> getMedicinesByApplicationArea(String areaName) {
        return medicineRepository.findByApplicationArea(areaName);
    }

    public List<Medicine> getMedicinesByType(String typeName) {
        return medicineRepository.findByTypeName(typeName);
    }
}

File: frontend/src/services/medicineService.ts
import { api } from './api';
import { Medicine, CreateMedicineRequest, MedicineType, ApplicationArea } from '../types/medicine';

export const medicineService = {
    async getAllMedicines(): Promise<Medicine[]> {
        const response = await api.get('/api/medicines');
        return response.data;
    },

    async getMedicineById(id: number): Promise<Medicine> {
        const response = await api.get(`/api/medicines/${id}`);
        return response.data;
    },

    async createMedicine(medicineData: CreateMedicineRequest): Promise<Medicine> {
        const response = await api.post('/api/medicines', medicineData);
        return response.data;
    },

    async updateMedicine(id: number, medicineData: Partial<Medicine>): Promise<Medicine> {
        const response = await api.put(`/api/medicines/${id}`, medicineData);
        return response.data;
    },

    async deleteMedicine(id: number): Promise<void> {
        await api.delete(`/api/medicines/${id}`);
    },

    async searchMedicines(name: string): Promise<Medicine[]> {
        const response = await api.get(`/api/medicines/search?name=${name}`);
        return response.data;
    },

    async getExpiredMedicines(): Promise<Medicine[]> {
        const response = await api.get('/api/medicines/expired');
        return response.data;
    },

    async getExpiringSoonMedicines(days: number = 7): Promise<Medicine[]> {
        const response = await api.get(`/api/medicines/expiring-soon?days=${days}`);
        return response.data;
    },

    async getMedicineTypes(): Promise<MedicineType[]> {
        const response = await api.get('/api/medicine-types');
        return response.data;
    },

    async getApplicationAreas(): Promise<ApplicationArea[]> {
        const response = await api.get('/api/application-areas');
        return response.data;
    }
};